recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,384 @@
++  bicep?: BicepConfigProperties;
+
+  @doc("Environment variables injected during Terraform Recipe execution for the recipes in the environment.")
+  env?: EnvironmentVariables;
+}
+
++@doc(Configuration for Bicep Recipes. Controls how Bicep plans and applies templates as part of Recipe deployment.)
++model BicepConfigProperties {
++
++  @doc("Authentication information used to access private bicep registries, which is a map of registry hostname to secret config that contains credential information.")
++  authentication?: Record<RegistrySecretConfig>;
+}
+```
+`RegistrySecretConfig` contains registry credential information, we also want users to specify the type of credentials, e.g. username-password, federated credentials.
+
+Option 1: 
+
+Adding a `type` property to `RegistrySecretConfig`
+environment.tsp
+```
+@doc(RegistrySecretType represents the authentication type used to authenticate private Bicep registries.)
+enum RegistryAuthType {
+  @doc("basicAuthentication type is used to represent username and password based authentication and the secretstore resource is expected to have the keys 'username' and 'password'")
+  basicAuthentication
+
+  @doc("azureFederatedIdentity type is used to represent registry authentication using azure federated identity and the secretstore resource is expected to have the keys 'clientId' and 'tenantId'")
+  azureFederatedIdentity
+
+  @doc("awsIRSA type is used to represent registry authentication using AWS IRSA.")
+  awsIRSA
+}
+
+@doc("Registry Secret Configuration used to authenticate to private bicep registries.")
+model RegistrySecretConfig {
+  @doc("The ID of an Applications.Core/SecretStore resource containing credential information used to authenticate private container registry.The keys in the secretstore depends on the type.")
+  secret?: string;
+
+  @doc(RegistryAuthType represents the authentication type used to authenticate private Bicep registries.)
+  type: RegistryAuthType
+}
+```
+Option 2:
+
+Adding support to the `Applications.Core/secretStores` resource to include additional types. Today `secretStores` resource only supports types `certificate` and `generic`.
+
+secretstores.tsp
+
+```diff
+@doc("The type of SecretStore data")
+enum SecretStoreDataType {
+  @doc("Generic secret data type")
+  generic,
+
+  @doc("Certificate secret data type")
+  certificate,
+
++  @doc("basicAuthentication type is used to represent username and password based authentication and the secretstore resource is expected to have the keys 'username' and 'password'")
++  basicAuthentication
++
++  @doc("azureFederatedIdentity type is used to represent registry authentication using azure federated identity and the secretstore resource is expected to have the keys 'clientId' and 'tenantId'")
++  azureFederatedIdentity
++
++  @doc("awsIRSA type is used to represent registry authentication using AWS IRSA.")
++  awsIRSA
```

Comment:
Have we considered either making it more generic - use `federatedIdentity` for both aws and azure, or make it more specific `azureWorkloadIdentity` because that's what we are supporting. @willtsai @reshmarahim thoughts?

Reply:
I think making it more specific to the cloud providers like `azureWorkloadIdentity` and `awsIRSA` would be easier for the users to interpret. Also expanding IRSA to IAM Roles for Service accounts in product/code docs would be helpful.